lib/beacon/template/heex/json_encoder.ex

```diff
@@ -0,0 +1,228 @@
+defmodule Beacon.Template.HEEx.JSONEncoder do
+  @moduledoc false
+
+  alias Beacon.Template.HEEx.HEExDecoder
+
+  @type token :: map()
+
+  @doc """
+  Encodes a HEEx `template` into a format that can be encoded into JSON.
+
+  The returned data structure can be serialized and sent over the wire to any client that uses a JSON API,
+  and it aims to retain all the information we need to reconstruct back the original template from it.
+
+  ## Data Structure
+
+  The encoded data structured emitted at the end is a list of tokens componsed of either a `heex_node` or a `eex_node`,
+  as specified below:
+
+      tokens = [heex_node() | eex_node()]
+
+      heex_node = %{
+        "tag" => String.t(),
+        "attrs" => %{String.t() => String.t()},
+        "content" => content(),
+        "rendered_html" => String.t()
+      }
+
+      eex_node = %{
+       "tag" => "eex_block",
+       "arg" => String.t(),
+       "blocks" => [%{"key" => String.t(), "content" => content()}]
+      }
+
+      content = [heex_node() | eex_node() | String.t()]
+
+  Note that:
+
+    * `rendered_html` key is optional
+
+  ## Example
+
+      iex> encode(~S|
+      <header>
+        <.link patch="/">
+          <h1 class="bg-red">My Site</h1>
+        </.link>
+      </header>
+      |, :my_site)
+      [
+        %{
+          "attrs" => %{},
+          "content" => [
+            %{
+              "attrs" => %{"patch" => "/"},
+              "content" => [
+                %{
+                  "attrs" => %{"class" => "bg-red"},
+                  "content" => ["My Site"],
+                  "tag" => "h1"
+                }
+              ],
+              "rendered_html" => "<a href=\"/\" data-phx-link=\"patch\" data-phx-link-state=\"push\">\n    <h1 class=\"bg-red\">My Site</h1>\n  </a>",
+              "tag" => ".link"
+            }
+          ],
+          "tag" => "header"
+        }
+      ]
+
+  """
+  @spec encode(Beacon.Types.Site.t(), String.t(), map()) :: {:ok, [token()]} | {:error, String.t()}
+  def encode(site, template, assigns \\ %{}) when is_atom(site) and is_binary(template) and is_map(assigns) do
+    case Beacon.Template.HEEx.Tokenizer.tokenize(template) do
+      {:ok, tokens} -> {:ok, encode_tokens(tokens, site, assigns)}
```

Comment:
Makes sense, we don't care about the intermediate AST representation, only about the final form.